Engine/EffectInstanceRenderRoI.cpp

```diff
@@ -736,39 +728,44 @@ EffectInstance::renderRoI(const RenderRoIArgs & args,
     ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
     ////////////////////////////// Compute RoI depending on render scale ///////////////////////////////////////////////////
 
-
     RectI downscaledImageBoundsNc = rod.toPixelEnclosing(args.mipMapLevel, par);
     RectI upscaledImageBoundsNc = rod.toPixelEnclosing(0, par);
-    {
-        ///Make sure the RoI falls within the image bounds
-        ///Intersection will be in pixel coordinates
-        if (frameArgs->tilesSupported) {
-            if (renderFullScaleThenDownscale) {
-                if ( !roi.clipIfOverlaps(upscaledImageBoundsNc) ) {
-                    return eRenderRoIRetCodeOk;
-                }
-                assert(roi.x1 >= upscaledImageBoundsNc.x1 && roi.y1 >= upscaledImageBoundsNc.y1 &&
-                       roi.x2 <= upscaledImageBoundsNc.x2 && roi.y2 <= upscaledImageBoundsNc.y2);
-            } else {
-                if ( !roi.clipIfOverlaps(downscaledImageBoundsNc) ) {
-                    return eRenderRoIRetCodeOk;
-                }
-                assert(roi.x1 >= downscaledImageBoundsNc.x1 && roi.y1 >= downscaledImageBoundsNc.y1 &&
-                       roi.x2 <= downscaledImageBoundsNc.x2 && roi.y2 <= downscaledImageBoundsNc.y2);
-            }
-#ifndef NATRON_ALWAYS_ALLOCATE_FULL_IMAGE_BOUNDS
-            ///just allocate the roi
-            upscaledImageBoundsNc.clipIfOverlaps(roi);
-            downscaledImageBoundsNc.clipIfOverlaps(args.roi);
-#endif
+    RectI roi;
+
+    if (renderFullScaleThenDownscale) {
+        //We cache 'image', hence the RoI should be expressed in its coordinates
+        //renderRoIInternal should check the bitmap of 'image' and not downscaledImage!
+        roi = args.roi.toNewMipMapLevel(args.mipMapLevel, 0, par, rod);
+
+        if (frameArgs->tilesSupported && !roi.clipIfOverlaps(upscaledImageBoundsNc)) {
```

Comment:
I don't think it's good practice to have the condition actually modify the context. I think we should move all clipIfOverlaps() calls out of the conditions.
This condition would then look like:
```
if (frameArgs->tilesSupported) {
  auto clipped = roi.clipIfOverlaps(upscaledImageBoundsNc);
  if (!clipped) {
     return eRenderRoIRetCodeOk;
   }
}
```
but I'm not sure that's the intent here, because a few lines below there's `const RectI originalRoI = roi;` but `roi`is not the original but the clipped roi... what do you think?

Reply:
If roi was used elsewhere in the conditional I might agree, but it isn't so I disagree with the proposed substitute. In my opinion the proposal is unnecessarily verbose and it assigns a name to a value that really is only needed for the conditional that follows. In my experience, assigning names to values is usually a signal that they will be used more than once, but in cases like this it is not true which is why I don't really see the need for giving it a name. 

Unfortunately I don't know the history of why the variable below is called originalRoI, but I suspect it is intended to store the value of roi before any further modifications occur after that point. I'm not sure what a better name might be given the complexity of this function and how clipping is applied(or not) based on a variety of conditions. I too was initially confused by the use of the term 'original' given all the possible roi transformations that occur before that declaration.

Reply:
I think side effects in conditions are bad practice in general (except for some constructs where the language instructs you to use side effects like for or while), because a condition should check the state, not modify it. This makes the code error-prone and harder to read.